File: autocapture/quality.py
"""Per-frame quality scoring (pure numpy).

Score is in [0, 1] — higher is better. The capture gate fires when
score >= ``cfg.capture_threshold`` AND the tracker has been stable for
N consecutive frames.
"""

from __future__ import annotations
import numpy as np
from .detector.quad import quad_area, quad_aspect
from .dewarp.shading import gaussian_blur


def _laplacian_variance(gray: np.ndarray) -> float:
    """Variance of the discrete Laplacian — a classic sharpness measure."""
    g = gray.astype(np.float64)
    p = np.pad(g, 1, mode="edge")
    lap = (
        -4 * p[1:-1, 1:-1]
        + p[:-2, 1:-1] + p[2:, 1:-1]
        + p[1:-1, :-2] + p[1:-1, 2:]
    )
    return float(lap.var())


def _sobel_mag_mean(gray: np.ndarray) -> float:
    """Mean Sobel gradient magnitude on a 2D uint8 image."""
    g = gray.astype(np.float64)
    p = np.pad(g, 1, mode="edge")
    # gx: derivative in x
    gx = (
        -p[:-2, 1:-1] + p[2:, 1:-1]
        - 2 * p[1:-1, :-2] + 2 * p[1:-1, 2:]
        - p[:-2, 2:] + p[2:, 2:]
    )
    # gy: derivative in y (same kernel, applied along rows)
    gy = (
        -p[1:-1, :-2] + p[1:-1, 2:]
        - 2 * p[:-2, 1:-1] + 2 * p[2:, 1:-1]
        - p[2:, :-2] + p[2:, 2:]
    )
    return float(np.sqrt(gx * gx + gy * gy).mean())


def _glare_ratio(gray: np.ndarray, quad: np.ndarray) -> float:
    """Fraction of pixels inside the quad that are clipped to white."""
    from .dewarp.shading import make_binary_mask
    mask = make_binary_mask(gray.shape, quad)
    if mask.sum() == 0:
        return 0.0
    over = (gray > 240) & (mask > 0)
    return float(over.sum()) / float(mask.sum())


def score(quad: np.ndarray, frame_bgr: np.ndarray, cfg) -> float:
    """Compute the per-frame quality score in [0, 1]."""
    h, w = frame_bgr.shape[:2]
    if frame_bgr.ndim == 3:
        gray = frame_bgr.mean(axis=2)
    else:
        gray = frame_bgr

    # 1. area
    frame_area = h * w
    area = quad_area(quad)
    area_ratio = area / frame_area
    if area_ratio < cfg.min_area_ratio:
        return 0.0
    area_score = min(area_ratio / cfg.target_area_ratio, 1.0)

    # 2. straightness — ratio of min to max bounding edge
    q = quad.astype(np.float64)
    edges = []
    for i in range(4):
        edges.append(np.linalg.norm(q[i] - q[(i + 1) % 4]))
    e_min, e_max = min(edges), max(edges)
    straightness = (e_min / e_max) if e_max > 0 else 0.0

    # 3. sharpness (Laplacian variance)
    lap_var = _laplacian_variance(gray)
    sharpness = min(lap_var / cfg.sharpness_full, 1.0)

    # 4. exposure (mean luma in the quad region)
    from .dewarp.shading import make_binary_mask
    mask = make_binary_mask(gray.shape, quad)
    if mask.sum() > 0:
        luma_mean = float(gray[mask > 0].mean())
    else:
        luma_mean = float(gray.mean())
    if cfg.exposure_low < luma_mean < cfg.exposure_high:
        exposure = 1.0
    elif luma_mean <= 0 or luma_mean >= 255:
        exposure = 0.0
    else:
        # smooth penalty outside the band
        if luma_mean < cfg.exposure_low:
            exposure = max(0.0, luma_mean / cfg.exposure_low)
        else:
            exposure = max(0.0, (255 - luma_mean) / (255 - cfg.exposure_high))

    # 5. blur (gradient magnitude)
    blur_mean = _sobel_mag_mean(gray)
    blur = min(blur_mean / cfg.blur_full, 1.0)

    score = (
        cfg.weight_area * area_score +
        cfg.weight_straightness * straightness +
        cfg.weight_sharpness * sharpness +
        cfg.weight_exposure * exposure +
        cfg.weight_blur * blur
    )
    return float(np.clip(score, 0.0, 1.0))

File: src/quality.js
// Per-frame quality scoring.
//
// Direct port of autocapture/quality.py.
//
// Score is in [0, 1] — higher is better. Components:
//   area, straightness, sharpness (Laplacian variance),
//   exposure (mean luma in mask), blur (Sobel magnitude mean).

import { quadArea } from "./detector/quad.js";
import { makeBinaryMask } from "./dewarp/shading.js";

function laplacianVariance(gray, H, W) {
  // 4-neighbor Laplacian: -4*p + up + down + left + right
  let sum = 0, sumSq = 0;
  for (let i = 1; i < H - 1; i++) {
    for (let j = 1; j < W - 1; j++) {
      const c = gray[i * W + j];
      const lap = -4 * c
        + gray[(i - 1) * W + j]
        + gray[(i + 1) * W + j]
        + gray[i * W + (j - 1)]
        + gray[i * W + (j + 1)];
      sum += lap;
      sumSq += lap * lap;
    }
  }
  const n = (H - 2) * (W - 2);
  const mean = sum / n;
  return sumSq / n - mean * mean;
}

function sobelMagMean(gray, H, W) {
  // Mean Sobel magnitude using 3x3 kernels. Same approach as
  // detector/classical.js sobelMag but we only need the mean, not the
  // full image, so we don't allocate the output array.
  const KX = [-1, 0, 1, -2, 0, 2, -1, 0, 1];
  const KY = [-1, -2, -1, 0, 0, 0, 1, 2, 1];
  let sum = 0, count = 0;
  for (let i = 1; i < H - 1; i++) {
    for (let j = 1; j < W - 1; j++) {
      let gx = 0, gy = 0;
      for (let ki = -1; ki <= 1; ki++) {
        for (let kj = -1; kj <= 1; kj++) {
          const v = gray[(i + ki) * W + (j + kj)];
          gx += KX[(ki + 1) * 3 + (kj + 1)] * v;
          gy += KY[(ki + 1) * 3 + (kj + 1)] * v;
        }
      }
      sum += Math.sqrt(gx * gx + gy * gy);
      count++;
    }
  }
  return count > 0 ? sum / count : 0;
}

/**
 * @param {number[]} quad flat 8-array, ordered TL TR BR BL
 * @param {Uint8ClampedArray|Uint8Array} frameBGR length H*W*3 or H*W*4 (RGBA)
 * @param {number} H
 * @param {number} W
 * @param {object} cfg quality config
 * @returns {number} score in [0, 1]
 */
export function score(quad, frameBGR, H, W, cfg) {
  // Convert to grayscale if needed
  const isRGBA = frameBGR.length === H * W * 4;
  const C = isRGBA ? 4 : 3;
  const gray = new Float64Array(H * W);
  for (let i = 0; i < H * W; i++) {
    let s;
    if (isRGBA) {
      s = (frameBGR[i * 4] + frameBGR[i * 4 + 1] + frameBGR[i * 4 + 2]) / 3.0;
    } else {
      s = (frameBGR[i * 3] + frameBGR[i * 3 + 1] + frameBGR[i * 3 + 2]) / 3.0;
    }
    gray[i] = s;
  }
  void C;

  // 1. area
  const frameArea = H * W;
  const area = quadArea(quad);
  const areaRatio = area / frameArea;
  if (areaRatio < cfg.minAreaRatio) return 0.0;
  const areaScore = Math.min(areaRatio / cfg.targetAreaRatio, 1.0);

  // 2. straightness
  const edges = [
    dist(quad[0], quad[1], quad[2], quad[3]),
    dist(quad[2], quad[3], quad[4], quad[5]),
    dist(quad[4], quad[5], quad[6], quad[7]),
    dist(quad[6], quad[7], quad[0], quad[1]),
  ];
  const eMin = Math.min(...edges), eMax = Math.max(...edges);
  const straightness = eMax > 0 ? eMin / eMax : 0.0;

  // 3. sharpness
  const lapVar = laplacianVariance(gray, H, W);
  const sharpness = Math.min(lapVar / cfg.sharpnessFull, 1.0);

  // 4. exposure (mean luma in the quad region)
  const mask = makeBinaryMask([H, W], quad);
  let maskSum = 0, lumaSum = 0;
  for (let i = 0; i < H * W; i++) {
    if (mask[i]) { lumaSum += gray[i]; maskSum++; }
  }
  const lumaMean = maskSum > 0 ? lumaSum / maskSum : gray.reduce((s, v) => s + v, 0) / gray.length;

  let exposure = 1.0;
  if (cfg.exposureLow < lumaMean && lumaMean < cfg.exposureHigh) {
    exposure = 1.0;
  } else if (lumaMean <= 0 || lumaMean >= 255) {
    exposure = 0.0;
  } else if (lumaMean < cfg.exposureLow) {
    exposure = Math.max(0, lumaMean / cfg.exposureLow);
  } else {
    exposure = Math.max(0, (255 - lumaMean) / (255 - cfg.exposureHigh));
  }

  // 5. blur
  const blurMean = sobelMagMean(gray, H, W);
  const blur = Math.min(blurMean / cfg.blurFull, 1.0);

  const total = cfg.weightArea * areaScore
              + cfg.weightStraightness * straightness
              + cfg.weightSharpness * sharpness
              + cfg.weightExposure * exposure
              + cfg.weightBlur * blur;
  return Math.max(0, Math.min(1, total));
}

function dist(x1, y1, x2, y2) {
  const dx = x1 - x2, dy = y1 - y2;
  return Math.sqrt(dx * dx + dy * dy);
}